### PR TITLE
Unify system-contracts migration and staged-contracts migration

### PR DESCRIPTION
**Pull request by SupunS**
Work towards https://github.com/onflow/cadence/issues/2865

Depends on https://github.com/onflow/cadence/pull/3117

Validator is disabled by default.

I also realized the core-contract migration and the staged contract migration contains a lot of duplicate logic.
Hence refactored the core-contracts migration (`ChangeContractCodeMigration`) to also use the `StagedContractMigration` underneath. Core contract changes are also in a way, a set of staged contract changes.

### CODE REVIEW

File: cmd/util/ledger/migrations/staged_contracts_migration.go [hunk 1]
```diff
@@ -144,10 +171,17 @@ func (m *StagedContractsMigration) MigrateAccount(
...
+
 	config := util.RuntimeInterfaceConfig{
 		GetContractCodeFunc: func(location runtime.Location) ([]byte, error) {
 			return m.contractsByLocation[location], nil
 		},
+		GetOrLoadProgramListener: func(location runtime.Location, program *interpreter.Program, err error) {
+			if err == nil {
+				elaborations[location] = program.Elaboration
+			}
+		},
```
Comment (SupunS) [inline cmd/util/ledger/migrations/staged_contracts_migration.go:184]: Is this safe? or is there a better way to collect the list of elaborations?

File: go.mod [hunk 2]
```diff
@@ -51,7 +51,7 @@ require (
 	github.com/multiformats/go-multiaddr-dns v0.3.1
 	github.com/multiformats/go-multihash v0.2.3
 	github.com/onflow/atree v0.6.1-0.20230711151834-86040b30171f
-	github.com/onflow/cadence v1.0.0-M7
+	github.com/onflow/cadence v1.0.0-preview.2.0.20240216110422-a67dbbf21171
```
Comment (turbolent) [inline go.mod:54]: This refers to a commit that isn't on any branch. 

It looks like this PR depends on https://github.com/onflow/cadence/pull/3111

File: cmd/util/ledger/migrations/staged_contracts_migration.go (hunk 1 above)
Comment (turbolent) [reply cmd/util/ledger/migrations/staged_contracts_migration.go:184]: This is probably fine

**Review by turbolent — COMMENTED** [review]
Thank you for refactoring the duplicated code!

Does this mean system contract updates are now going to be checked by the updated contract update checker, like other staged contract updates, and forced to pass?

File: cmd/util/ledger/migrations/change_contract_code_migration.go [hunk 3]
```diff
@@ -1,175 +1,46 @@
...
-		d.contracts = map[common.Address]map[flow.RegisterID]string{}
-	}
-
-	if _, ok := d.contracts[address]; !ok {
-		d.contracts[address] = map[flow.RegisterID]string{}
+func NewChangeContractCodeMigration(chainID flow.ChainID) *ChangeContractCodeMigration {
+	return &ChangeContractCodeMigration{
+		StagedContractsMigration: NewStagedContractsMigration(chainID).
+			// TODO:
+			//WithContractUpdateValidation().
```
Comment (SupunS) [inline cmd/util/ledger/migrations/change_contract_code_migration.go:29]: > Does this mean system contract updates are now going to be checked by the updated contract update checker, like other staged contract updates, and forced to pass?

@turbolent It is not - buy default. We can enable it here if we want.

File: go.mod (hunk 2 above)
Comment (SupunS) [reply go.mod:54]: yeah, updated the description

File: cmd/util/ledger/migrations/change_contract_code_migration.go (hunk 3 above)
Comment (turbolent) [reply cmd/util/ledger/migrations/change_contract_code_migration.go:29]: How can we disable it for system contracts and enable it for user contracts?

**Review by turbolent — APPROVED** [review]
Great work refactoring and cleaning this up!

File: cmd/util/ledger/migrations/change_contract_code_migration.go (hunk 3 above)
Comment (SupunS) [reply cmd/util/ledger/migrations/change_contract_code_migration.go:29]: It is also disabled in this PR. Enabled it in a follow-up PR: https://github.com/onflow/flow-go/pull/5415/files#diff-efc0583bcfd2aa19725211a31ed4ceed33b7d0ab5228c4ab533acf0d7f474d63R201-R202.

Comment (SupunS) [reply cmd/util/ledger/migrations/change_contract_code_migration.go:29]: I can do that change in this PR itself if that's better

Comment (turbolent) [reply cmd/util/ledger/migrations/change_contract_code_migration.go:29]: Nice! So with https://github.com/onflow/flow-go/pull/5415/files#diff-efc0583bcfd2aa19725211a31ed4ceed33b7d0ab5228c4ab533acf0d7f474d63R201-R202, now all contract updates, both "system contracts" and "user contracts", will get validated?

Comment (SupunS) [reply cmd/util/ledger/migrations/change_contract_code_migration.go:29]: so basically, we can enable/disable it in the corresponding constructor `NewCadence1ContractsMigrations` and `NewChangeContractCodeMigration`

Comment (turbolent) [reply cmd/util/ledger/migrations/change_contract_code_migration.go:29]: Great! Thank you for the explanation

Comment (SupunS) [reply cmd/util/ledger/migrations/change_contract_code_migration.go:29]: > now all contract updates, both "system contracts" and "user contracts", will get validated?

only the user contracts get validated for now. Validation for system contracts still remains disabled.

Comment (SupunS) [reply cmd/util/ledger/migrations/change_contract_code_migration.go:29]: Sorry, the github isn't showing (your) comments until I refresh and I replied without refreshing 😅

**Pull request merged**